File: tests/xcalablemp/NOT-IMPLEMENTED/59/reduction_async4.c
/*testp033.c*/
/*reduction指示文(&)のテスト*/
#include<mpi.h>
#include<xmp.h>
#include<stdio.h>
#include<stdlib.h>   
#pragma xmp nodes p(*)
int procs,id;
int mask1,val1;
int mask2,val2;
char *result;
int i,w;
int main(void){

   if(xmp_num_nodes() > 31){
      printf("%s\n","You have to run this program by less than 32 nodes.");
   }

   procs = xmp_num_nodes();
   id = xmp_num_nodes()-1;
   result = "OK";
   w=1;
   for(i=0;i<procs;i++){
      w*2;
   }
   for(i=0;i<w;i=i+2){
      mask1 = 1 << id;
      val1 =!(i & mask1);
#pragma xmp reduction(&:val1) async(1)
      mask2 = 1 << id;
      val2 =!((i+1) & mask2);
#pragma xmp reduction(&:val2) async(2)
#pragma xmp wait_async(1)
      if((!(val1)) != i){
         result = "NG";
      }
#pragma xmp wait_async(2)
      if((!(val2)) != i+1){
         result = "NG";
      }
   }

   printf("%d %s %s\n",xmp_node_num(),"testp033.c",result); 
   return 0;
}
      
         
      
   

File: tests/XMP/NOT-IMPLEMENTED/59/reduction_async4.f90
! testp033.f
! reduction指示文(iand)のテスト

      program main
      include 'xmp_lib.h'
!$xmp nodes p(*)
      integer procs, id
      integer mask1, val1
      integer mask2, val2
      character(len=2) result

      if(xmp_num_nodes().gt.31) then
         print *, 'You have to run this program by less than 32 nodes.'
      endif


      procs = xmp_num_nodes()
      id = xmp_node_num()-1
      result = 'OK'
      do i=0, 2**procs-1, 2
         mask1 = lshift(1, id)
         val1 = not(iand(i, mask1))
!$xmp reduction(iand: val1) async(1)
         mask2 = lshift(1, id)
         val2 = not(iand(i+1, mask2))
!$xmp reduction(iand: val2) async(2)
!$xmp wait_async(1)
         if(not(val1) .ne. i) then
            result = 'NG'
         endif
!$xmp wait_async(2)
         if(not(val2) .ne. i+1) then
            result = 'NG'
         endif
      enddo

      print *, xmp_node_num(), 'testp033.f ', result

      end
